### ISSUE TITLE
Help with using parts of scripts to accomplish fewer things at once

### ISSUE THREAD

**Issue by jrochet** [OfficeDev/Office-IT-Pro-Deployment-Scripts#219]
Hello,

Just wondering if there is anywhere for people to get help using these scripts.  There is a lot to them and each one seems to be an entire use case, end to end.  I would like to be able to piece a few of the scripts together to do fewer things but they all seem to start at downloading the files, just as an example.  I'd I've already downloaded the files to a share, and now just want to have a good startup script install click to run, only if it's not installed already, and then have another startup script to uninstall older versions, is this something that is possible?  The configure-gpo**"" script is seemingly only ready to do the entire solution from scratch every time.

Any help would be greatly appreciated.  The readme's are good, but I guess having read them has led me to these questions.

Thanks again.

**Comment by jrochet** [comment]
I should say, mainly if there was some assistance getting Remove-PreviousOfficeInstalls.ps1 to work as a machine startup script, that would be great.  The way the function within that script "GetScriptPath" works, I was thinking that I should be able to just put Remove-PreviousOfficeInstalls.ps1 and the various OffScrub VBS's into my GPO's folder \\domain.com\sysvol\domain.com\policies\{Guid}\Machine\Scripts\Startup.  But that doesn't work, and what I'm trying is to have the script as a PowerShell startup script with the arguments:
-RemoveClickToRunVersions $false -Remove2016Installs $true -KeepUserSettings $true -ProductsToRemove MainOfficeProduct

However in testing that fails.  If I manually try to run the script from there, GetScriptPath fails to set the path to the GPO UNC, and instead just sets the path to be my working directory.

**Comment by gregbesso** [comment]
Hey there jrochet, I'm just starting to use this script also. I'm having some success, can you let us know if when you run the script what the failure/error is? Is there any message perhaps, such as the execution policy restrictions being hit? I run my PS commands bypassing those to ensure workstation settings to prevent it from running. I could run an example by you once you verify the issue you're having...

Thanks,
Greg

**Comment by jrochet** [comment]
Hi Greg,

Well the latest is that I decided to use GPO computer configuration -> Preferences to put the files (Off*Scrub.vbs and Remove-PreviousOfficeInstalls.ps1) into %Windir%\Temp and then my GPO's startup script is just a helper script to launch the local Remove-PreviousOfficeInstalls.ps1 script.  At this point the script runs, and nothing happens.  If I run this helper script manually from an administrator PowerShell, I see "Detecting office versions..." and then soon after it exists.  It's not detecting my Office install (2007 pro plus).  here's is my working helper script for use as a GPO machine startup script:

```
$VbsFilesExist = Get-ChildItem -Path $env:windir\Temp -Filter OffScrub*.vbs
$Timer = 1
do {
  Write-Host "Waiting for VBS files to copy to %Windir%\Temp."
  Start-Sleep -Seconds 5
  $VbsFilesExist = Get-ChildItem -Path $env:windir\Temp -Filter OffScrub*.vbs
  $Timer++
}
while (-not ($VbsFilesExist) -and ($Timer -le 12))

$Ps1FileExists = Get-ChildItem -Path $env:windir\Temp -Filter Remove-PreviousOfficeInstalls.ps1
$Timer = 1
do {
  Write-Host "Waiting for PS1 file to copy to %Windir%\Temp."
  Start-Sleep -Seconds 5
  $Ps1FileExists = Get-ChildItem -Path $env:windir\Temp -Filter Remove-PreviousOfficeInstalls.ps1
  $Timer++
}
while (-not ($Ps1FileExists) -and ($Timer -le 12))

$ExecutionPolicy = Get-ExecutionPolicy
Set-ExecutionPolicy RemoteSigned -Force

& $($Ps1FileExists.FullName) -RemoveClickToRunVersions $false -Remove2016Installs $true -KeepUserSettings $true -ProductsToRemove MainOfficeProduct

Set-ExecutionPolicy $ExecutionPolicy -Force

```

Now onto just troubleshooting the Remove-PreviousOfficeInstalls.ps1 script and once I can make it work locally, it should begin to work with my GPO solution.  I wish the script did some logging to help me out, it's a lot of reverse engineering to figure it out.

**Comment by gregbesso** [comment]
Ah ok. So the things I've noticed are that if I run it manually I have to do it in an ISE or PowerShell session as administrator. Also n some older Windows 7 computers, I also am going to have to copy the VBS files to C:\Windows\System32 (don't ask me why, but I found it got around some issues on those systems). I'm just wrapping my call to the UNC based Remove-PreviousOfficeInstalls with a Copy-Item command to get those files in place before calling it. 

Also some of the parameters have suitable values by default so you don't even need to pass them (such as KeepUserSettings).

So, I'm dot sourcing the ps1 file via UNC path directly, not copying it locally.

What OS are you on? Just curious...

**Comment by jrochet** [comment]
I'm testing with Windows 7 SP1 64-bit Enterprise N.  The "N" I don't know if that will make a difference.  I do have an update.  I've found that the script "Remove-PreviousOfficeInstalls.ps1" itself is part of the problem.

I've found that in order to remove Office 2007 (will be testing 2010 today as well), I have to either run it with no parameters (.\Remove-PreviousOfficeInstalls.ps1), or ironically, if I include "-WhatIf" it still uninstalls everything.  So I have tested this both manual and locally, and also have tested by updating my helper script to be like this at the end:

```
$ExecutionPolicy = Get-ExecutionPolicy
Set-ExecutionPolicy RemoteSigned -Force

& $($Ps1FileExists.FullName)

Set-ExecutionPolicy $ExecutionPolicy -Force
```
Note the "& $($Ps1FileExists.FullName)" line is just calling the script, not adding any parameters.  This works.  I believe the logic in the script for handling the parameters is either containing errors, or is meant to only handle specific combinations of parameters.  The fact that the -WhatIf has no effect makes me think the logic in the script has some problems.  I'll let you know how it goes for Office 2010 (pro Plus as well).

**Comment by AStigall** [comment]
Hi @jrochet, I'd be more than happy to help you get the scripts working in your environment. To run the scripts from a GPO I think we'll need to specify the a path to the folder where your scripts are then setting the location to that path before running the Remove-PreviousOfficeInstalls script. 

I will put together a new script for you that should work with what you're trying to do. Then, when you create the GPO to run the script we can pass through a parameter for the scriptpath.

**Comment by jrochet** [comment]
Hi @AStigall (retro-mention to @gregbesso as well as I forgot).  Thanks for helping me out here!  I like the idea of changing the scriptpath so that the single script, Remove-PreviousOfficeInstalls.ps1, can be the GPO machine startup script, and the VBS files can be stored with the GPO too.  It's nice and clean that way.

I have created my own repo too for my workspace around this stuff - https://github.com/jrochet/OfficeDeployment

Let me know if taking copies and putting them in my repo is anything short of perfectly ok.

I think I will submit a new Bug for the issue that is how the Remove-PreviousOfficeInstalls.ps1 parameters are handled.

**Comment by AStigall** [comment]
@jrochet I have fixed the issue with the script not working when you specify a ProductToRemove. I will post the fixes and a new script for you shortly.

**Comment by letsgoflyers81** [comment]
@AStigall I have a somewhat related question about specifying which product to remove.  I need a script to remove all previous versions of a specific product, e.g. only Visio, only Project, or only Office Professional Plus.  Will the new script you're working on be able to do this?

**Comment by gregbesso** [comment]
Also I just gave up this morning trying to use OffScrub to remove existing ClickToRun builds. It was not really working. But luckily, the native uninstall using Config.xml and <Remove All="True"> and using Setup.exe calling it works wonderfully. In case useful figured would mention :-)

**Comment by lgusers** [comment]
So I'm using a customized version of 10-ExampleRemovePreviousAndUpgrade.ps1 to remove Office 2010 and Skype For Business 2016. I setup a GPO to make it a startup script, and my problem is that it works to well. Every reboot Office is uninstalled and then reinstalled. I'm thinking some kind of command to see if Office 2010 is installed, and if not, stop processing the script, but I'm struggling with what commands to use to accomplish this. Anyone have any suggestion on how I could do that, or how they did something similar?

**Comment by AStigall** [comment]
@letsgoflyers81 Yes, use the Remove-PreviousOfficeVersions.ps1 script and run the script using the -ProductsToRemove parameter. For example, if I want to only remove Microsoft Office Visio Professional 2007 I would enter: **Remove-PreviousOfficeVersions -ProductsToRemove Visio**

**Comment by AStigall** [comment]
@jrochet @gregbesso I've updated the GPO scripts. I found and fixed an issue when using MainOfficeProduct. I also added a a logging method that I hope you find helping when tracking the progress of the script. If you try it and feel it needs to be more granular or how it could be more uselful please let me know.

**Comment by AStigall** [comment]
I'm updating the README file with updated instructions.

**Comment by AStigall** [comment]
@lgusers Let's consider a couple options: 1) Add a check to see if Office 2010 is installed, if not then don't run. 2) Add a check to see if Office 365 ProPlus is installed. If yes, then don't run.

I'm always inclined to check for the upgraded version, like Office 365 ProPlus, instead of 2010 by the chance there's a case where someone has 2007 or 2013.

**Comment by lgusers** [comment]
Yah I think either option would work, I was hesitant to have it check on the newer version because everyone has Skype for business Basic 2016 installed, so if we just checked for 2016 it would trigger anyway. So I assume that would go at the beginning of the deployment script, but I'm struggling with what the command would look like, and how it would tell it end all functions.

**Comment by AStigall** [comment]
@lgusers You may be able to resolve this by changing line 36 to the following:

$Office2016C2RExists = $officeProducts | Where {$_.ClickToRun -eq $true -and $_.Version -like '16.*' -and $DisplayName -notmatch "Skype" }

I don't have a VM with Skype for Business Basic 2016 on it, but that seems like it should for you.

**Comment by jrochet** [comment]
Thanks very much @AStigall!  I will review it either tonight or tomorrow morning and let you know how it goes.  Double thanks for the logging addition.  This is great service:)!!

**Comment by jrochet** [comment]
Hey @AStigall .  I'm seeing your update for the OffScrub03.vbs 2013->2003 fix, but not seeing an updated version of Remove-PreviousOfficeInstalls.ps1.  I looked in both locations in both Master and Development branches.  Can you let me know where to find it?  I may have overlooked it accidentally.

**Comment by AStigall** [comment]
@jrochet I updated the copy of Remove-PreviousOfficeInstalls in the Setup-GPOOfficeDeployment folder. I will get the other locations updated this morning.

**Comment by jrochet** [comment]
Hi AStigall,

Sorry I didn't respond sooner. I have been testing with the updated Remove-PreviousOfficeInstalls.ps1 and confirmed that the -ProductsToRemove parameter is working as intended (only tested values AllOfficeProducts, MainOfficeProduct, and Visio).  I did however still have trouble calling the script remotely, particularly with Windows 7 SP1 Enterprise, with only PowerShell 2.0 installed.  The GetScriptPath function just wouldn't work.  I tried replacing the contents of that function with a few other ways to get the PSScriptRoot equivalent for PowerShell 2.0 (e.g. $MyInvocation.<other stuff>), but had no luck.. What I did, was to just set the $scriptPath variable at the top of the script to be the UNC path to the folder where I'm storing the script (\\domain.com\NETLOGON).  I've found it's easiest for me to store the .PS1's and the.VBS's in that location.  It's working exactly as needed and I'm able to repeat this for customers as needed.  So thanks very much.  I'm working on my own OneDrive next-generation sync client user logon script to prompt the setup now.  Thanks again.
